syntax = "proto3";

message TablesListRequest {}

message TablesListReply {
    repeated TableDefinition entries = 1;
}

message TableDefinition {
    enum TableType {
        UNSPEC = 0;
        HASH   = 1;
        ARRAY  = 2;
    }

    string table_name    = 1;
    TableType table_type = 2;
    uint32 key_size      = 3;
    uint32 value_size    = 4;
    uint32 max_entries   = 5;
}

message TableListRequest {
    string table_name = 1;
}

enum TableStatus {
    SUCCESS = 0;
    TABLE_NOT_FOUND = 1;
    ENTRY_NOT_FOUND = 2;
}

message TableListReply {
    TableStatus status = 1;
    TableDefinition entry = 2;
    uint32 n_items = 3;
    bytes items = 4;
}

message TableEntryGetRequest {
    string table_name = 1;
    bytes key = 2;
}

message TableEntryGetReply {
    TableStatus status = 1;
    bytes key = 2;
    bytes value = 3;
}

message TableEntryInsertRequest {
    string table_name = 1;
    bytes key = 2;
    bytes value = 3;
}

message TableEntryInsertReply {
    TableStatus status = 1;
}

message TableEntryDeleteRequest {
    string table_name = 1;
    bytes key = 2;
}

message TableEntryDeleteReply {
    TableStatus status = 1;
}
